### ISSUE TITLE
Calculus Ratiocinator vs. Characteristica Universalis

### ISSUE THREAD

**Issue by prathyvsh** [prathyvsh/history-of-logic#1]
* Algebra of Logic Tradition

The algebra of logic tradition can be thought of as starting from the time of Leibniz where he identifies two notions:
- Calculus Ratiocinator
- Characteristica Universalis

I have to identify who all before Boole worked on these traditions and how it evolved.

Also, I have to check and see if Boole had read Leibniz. (A cursory search shows that he came to know about Leibniz work 12 years after the publication of Laws of Thought)

**Comment by prathyvsh** [comment]
Logic as a Language vs. Logic as a Calculus idea is touched upon by Heijenoort.

Logic as a Calculus gives it a force whereby, the premises you put in gives your eventual deductions without addressing the epistemology of the premises.

Logic as a Language I think helps capture the world around us into linguistic forms.

There seems to be two more aspects of generation, and judgement in Leibniz’s work. I need to explore more to understand how those thrusts influenced further developments.

Both are not mutually exclusive notions. Leibniz’s lingua universalis had the idea of a calculus ratiocinator that would allow one to calculate the conclusions from the ideas represented. Also, interesting to note here is that Leibniz invented calculus as a part of his lingua universalis project.

**Comment by prathyvsh** [comment]
This paper by Jourdain discusses the ideas in some detail: https://github.com/prathyvsh/history-of-logic/blob/master/papers-read.org#the-logical-work-of-leibniz-1916

He aligns the work of Jevons, Peirce, Dedekind, Schröder, Hermann, Robert Grassmann, Hugh MacColl, John Venn, and many others in the line of calculus ratiocinator.

And the work of Russell, Whitehead, and Frege in the line of Characteristica Universalis dreamt by Leibniz.

I think such a hard delineation is probably not a good idea as Leibniz himself didn't see them as mutually exclusive.

**Comment by prathyvsh** [comment]
Papers on the Woodhouse/Herschel/Babbage/Peacock Analytic school:
https://royalsocietypublishing.org/doi/10.1098/rsnr.1990.0018
https://www.sciencedirect.com/science/article/pii/0315086080900038
https://www.maths.cam.ac.uk/opportunities/careers-for-mathematicians/summer-research-mathematics/files/Phillips.pdf

**Comment by sholtomaud** [comment]
> This paper by Jourdain discusses the ideas in some detail: https://github.com/prathyvsh/history-of-logic/blob/master/papers-read.org#the-logical-work-of-leibniz-1916
> 
> He aligns the work of Jevons, Peirce, Dedekind, Schröder, Hermann, Robert Grassmann, Hugh MacColl, John Venn, and many others in the line of calculus ratiocinator.
> 
> And the work of Russell, Whitehead, and Frege in the line of Characteristica Universalis dreamt by Leibniz.
> 
> I think such a hard delineation is probably not a good idea as Leibniz himself didn't see them as mutually exclusive.

Do you have some references that detail the specifics of Leibniz's  Characteristica Universalis?

**Comment by prathyvsh** [comment]
@sholtomaud In my understanding, this work is spread throughout the writing of Leibniz. So, I think the best way to understand it might be to start reading the original of Leibniz because they are inextricably interlinked. But as a started, may be these two papers might be a good starting point I think:

1/ Chapter 4 from Louis Couturat’s The Logic of Leibniz: http://philosophyfaculty.ucsd.edu/faculty/rutherford/Leibniz/Couturatchapters/Chap4.pdf
2/ Leibniz’ Logic by Lenzen: https://homepages.uc.edu/~martinj/Leibniz/W_Lenzen.pdf

Feel free to post here if you find some better reading material!

**Comment by sholtomaud** [comment]
Thanks @prathyvsh. Couturat's translated chapter is good, but doesn't really support anything substantial when I read it, so it opens up varying interpretations, but also leads to conflations between Characteristica and Calculus notation.

I used to refer to **Philosophical Papers and Letters. Edited by Leroy E Loemker. Second. Dordrecht : D. Reidel, 1970.** but again the Characteristica was sparse. 

The original Leibniz parchments are hard to get one's hands on, however the Bibliothek has more and more stuff available. I'm yet to scour for the Characteristica. http://digitale-sammlungen.gwlb.de/index.php?id=6&tx_dlf%5Bid%5D=2354&tx_dlf%5Bpage%5D=1

Couturat's Ch4, " where the thing itself lies far from the imagination or is too vast ...  may be subjected to the imagination by means of characters or shortcuts; and those things that cannot be depicted, such as intelligible entities, may nevertheless be depicted by a certain hieroglyphic, but at the same time philosophical, reason", takes me in a different direction to the Symbolic Logic folk, especially the mention of **hieroglyphics** and **Chinese images**, making it possibly more pictographic than lexical.

**Comment by prathyvsh** [comment]
@sholtomaud Yes, I have already sensed a lot of aberrations on how people interpret Leibniz’ work, which is why I stated that it is best to patch together straight from the source. Also, re: the difference, thought not fully convincing at places, I thought this was a decent read which links Characteristica Universalis and how it differs from Russell et al.’s tradition: https://monasandnomos.org/2012/12/05/the-idea-of-a-characteristica-universalis-between-leibniz-and-russell-and-its-relevancy-today/

Also, feel free to share if you have written anything on this. I am actively trying to research this niche as I think Leibniz got a lot of things (metaphysically) right to create a ground for geometric computing.

**Comment by sholtomaud** [comment]
Yeah there is an old paper Cevolatti and myself wrote [Realising the Enlightenment](https://www.sciencedirect.com/science/article/abs/pii/S0304380003005404) .  I would maintain the thesis but totally reverse the claims made therein about Cartesian philosophy. There are still two things that strike me:

* One is that Leibniz supposedly wanted to use the proceeds of his mine engineering consulting at the Hartz silver mines to develop the the CU. 
* The other is that, if we grant the argument that the Characters were intended to refer to pictograms, we can come at the existing modern Leibniz CU literature from another angle. This is to say that the connection to computation is made if we imagine the CR-CU pair intended to inform a method for constructing a simulation program with pictograms. 

If granted license to roam, we can see  this type of "computation with pictures" in the General Ecological Engineering Systems (GEES) literature.
![image](https://user-images.githubusercontent.com/398402/147848393-f94fbcf8-847e-4d14-9266-d12dafad2704.png)
See also [Modelling for all Scales](https://books.google.com.au/books?hl=en&lr=&id=PXNWwRXZU7cC&oi=fnd&pg=PP1&dq=odum+picture+math&ots=w7GkbWvRXe&sig=5q7NfSUua6uzlWCalHqZDGXNARc&redir_esc=y#v=onepage&q=odum%20picture%20math&f=false) 

What troubles is the gap between "picture math" and geometric computing I think you're referring to. 

Sure, what is pictured in the GEES method is loosely geometric - in as much as the models typically refer to geographic sites, and combine systems biology, hydrology, thermodynamics etc. - but it doesn't specifically refer to geometric transformation. 

One question then, where do you see Leibniz referring to geometric computing? Is it with reference to Characteristica?

**Comment by sholtomaud** [comment]
Also, when you use the term, 'geometric computing', do you mean 'computing with geometric forms', like, 'how does a cube compute?', or do you mean 'using a computer to generate a geometric form'?

**Comment by prathyvsh** [comment]
@sholtomaud That paper looks like a very interesting reference. Thanks for pointing me to it!

The sense in which I use the term geometric computing is actually a mix of both. The forms themselves act as substance and substrate to ground computation in. The one I have in mind is much close to what has been achieved with https://homotopy.io/

You manipulate geometrical forms and derive computational content out of it all the while maintaining a bidirectional link to the symbolic algebra which you use to generate these forms. You can relate how one thing morphs in relation with another like using a graph or a symplectic structure and this gives you the results that you want to get done. This way, both the medium in which the relations are cast are themselves geometric (topologic) objects in the global picture and objects used to enact the computation locally are also geometric. I only have vague intuitions on how to approach this at the moment, but I have been curating a thread on some of the mathematical works that tread in this direction here: https://twitter.com/prathyvsh/status/1286556976881627136

Especially look at the work of Rasetti / Merelli later down in the thread: https://twitter.com/prathyvsh/status/1322963506962984961

The part in which I intuited that this is what Leibniz wanted to pursue is through various parts of his work that I read from various places. But essentially, understanding that analysis situs (the first name of the field of topology), was a Leibnizian idea made me warm up to the idea that he had something derived from his metaphysics on how geometry of forms underly what he wanted to achieve with CU/CR. This talk by Maximilian Schich was also a good reference point on the analysis situs link: https://www.youtube.com/watch?v=dvlnDkfUq74

**Comment by sholtomaud** [comment]
@prathyvsh are you aware of Descartes' theorem?

**Comment by prathyvsh** [comment]
@sholtomaud No! How does it connect with geometric computation?

**Comment by sholtomaud** [comment]
It depends on how you define geo. comp. Taking your phrase, "manipulate geometrical forms and derive computational content". Do you mean that I have a substance which is some kind of electro-clay (eClay) that I can mould in such a way that even though it's the same substance in the same quantity, the different forms of the substance will perform different computations, and there is some way that we can inspect/plot the results of these computations?

As to Descartes' Theorem, it is also known as Euler's proof of the polyhedron formula; “In 1860, over a century after Euler presented his proof of the polyhedron formula, evidence surfaced that René Descartes, the famous philosopher, scientist, and mathematician, had known of this remarkable relationship in 1630” Excerpt From, Euler's Gem, David S. Richeson, p.165.

As to relevance: "Many of the currently available B-rep solid modeling systems represent objects by polyhedra, making the Descartes-Euler Theorem important in B-rep schemes." (See also Theorem 1, p. 764)
@article{jablokow1993topological,
  title={Topological and geometric consistency in boundary representations of solid models of mechanical components},
  author={Jablokow, AG and Uicker Jr, JJ and Turcic, DA},
  year={1993}
}

I've attached a representation of the theorem: 
![image](https://user-images.githubusercontent.com/398402/147894327-911d2856-abb8-4dd9-b1f3-ef0d8865b38e.png)

Can you see how this could be used in your geometric computation?

**Comment by prathyvsh** [comment]
@sholtomaud Yes, that is very much in line with what I am thinking of. But instead of clay, I started referring to it as bread dough. You knead it to a requisite shape and then derive computational content via it’s relationships, proportions, ratios with other objects in the environment. You can bake / unbake it and send it over wire for friends to play with, much in line with what is achieved with polynomials and geometry in Surfer: https://www.imaginary.org/program/surfer

Re: Descartes’ Theorem, I stumbled on this yesterday while googling that Euler got to this via Leibniz’ notebooks: http://www.math.stonybrook.edu/~tony/whatsnew/column/descartes-0899/descartes2.html

It looks like the book mentions figurate numbers and makes a connection with number theory. So it must be an important link to explore with respect to geometric computation. Thanks for mentioning it!

But correct me if I am wrong, but as mentioned in the Cultural Analysis Situs video outlined above, didn’t Poincaré after becoming aware of Leibniz’ work on topology elaborate on this idea of invariants of topological shapes to arbitrary shapes and dimensions with his work on [homology](https://www.quantamagazine.org/how-mathematicians-use-homology-to-make-sense-of-topology-20210511/) and [cohomology](http://blog.deinding.net/maths/cohomology.html)? Since I haven’t read the original is there anything that I should pay particular attention on something when reading Descartes’ work that is missing or is not covered in current direction of advances in topology?

**Comment by sholtomaud** [comment]
These are good questions. I'm here for the CU, but interested in the bread dough. I can't advise on Poincare unfortunately.

**Comment by prathyvsh** [comment]
@sholtomaud I don’t know how much in line with CU, topological computation is but since it is deeply connected with differential/integral calculus and number theory, which Leibniz pursued as well, topological knots might turn out to be the ”characters” in CU. Homology and Cohomology studies them as invariants which if you are familiar with [formal concept analysis](https://www.youtube.com/watch?v=Xuxm929tIRY) harks close to the idea of ”concepts”. I think there is some connection with intension/extension idea in Peirce/Leibniz etc. via knots as these things can be folded/unfolded. Because with concepts and categories we capture the ”invariance” about the thing being abstracted allowing it to be perceived from multiple perspectives and I think it is the same thing is happening in homology/cohomology. It is something I am furthering my research towards in the near future and I will be able to talk more once I have traversed the necessary breadth.

**Comment by sholtomaud** [comment]
So if I had had the capacity at hand in 1994 I would have proposed the following as a CU Character:
![image](https://user-images.githubusercontent.com/398402/148036537-fd363a43-b982-4b11-9655-d1cf876b4370.png)

It seems to me that the so-called conservation law was an invariant that had Leibniz's interest, and he needed a way to track the invariance in complex systems of interacting characters. 

Obviously any storage tank like that depicted above, has geometric features, but I'm not sure that those features (a knot say) was something that could have any immediate utility for Leibniz unless it could help him calculate the force that a tank could deliver over time.

Thoughts?

**Comment by prathyvsh** [comment]
@sholtomaud Try looking into that formal concept analysis video I shared. Though it gets at an idea of a concept via formal analysis, I am pretty positive that if that direction is developed further, we can arrive at topological ways to ground what a concept like a tank or its invariances governed by natural laws: stocks/flows etc. Graph based nodes and arrows are low dimensional topological objects. Homology and cohomology studies generalizations of these structures which can vary the arity between relations (think 3 things interacting together like in a hypergraph or 2 objects related in 2 different ways in a multigraph) using constructs like (co)chains, (co)cycles, and simplicial complexes. I feel that to complete/make progress on the Leibnizian program, a good course is by pursuing the great results and frameworks in advanced math because that is what Leibniz himself did. I see topology and (co)homology as natural outgrowth of what he pursued with calculus and analysis situs.

**Comment by sholtomaud** [comment]
I'll check out the vid. 

But yeah, if I've read the various patchy Leib. translations right, the CU was about simplifying the math for the everyperson. We want quick ways of creating new combinations that are energetically valid and so can be instantiated/fabricated. I've long had the question of whether there might be a way of relating the CU (as a pictographic model simulation framework) to CAD (as the platform for Cartesian geometry). Somehow you might get a UI that looks vaguely resembling this mangled cut&paste. where the top is the CU and bottom the CR:
![image](https://user-images.githubusercontent.com/398402/148113359-1116b4f2-1c07-4e44-82e7-f63528b09e11.png)

But the workflow here is left to right top to bottom:
![image](https://user-images.githubusercontent.com/398402/148114388-35cb07af-39fb-4c35-bc16-7b081f34e698.png)

whereas I think your method might be the reverse?

On the relevancy of Descartes' theorem for CAD; Jablokow et al make the interesting observation in the following:

![image](https://user-images.githubusercontent.com/398402/148112601-fe3ed31f-b75a-479a-9ddd-004c8db4ec34.png)


They go on to talk about the "Euler characteristic":

![image](https://user-images.githubusercontent.com/398402/148112726-efcf667a-2fe5-4fca-9a17-6457ff1716c6.png)
Jablokow, A. G., J. J. Uicker Jr., and D. A. Turcic. “Topological and Geometric Consistency in Boundary Representations of Solid Models of Mechanical Components.” Journal of Mechanical Design 115, no. 4 (December 1, 1993): 762–69. https://doi.org/10.1115/1.2919266.

**Comment by prathyvsh** [comment]
@sholtomaud > whereas I think your method might be the reverse?

I would have to report back on that once I work out how knots can be (co)algebraically composed/applied/folded/unfolded.

And thanks for the share on Descartes Theorem. Will see what it leads up to.
